lib/segment/src/index/field_index/numeric_index/mutable_numeric_index.rs

```diff
@@ -205,41 +234,68 @@ impl<T: Encodable + Numericable + Default> InMemoryNumericIndex<T> {
     }
 }
 
-impl<T: Encodable + Numericable + Default> MutableNumericIndex<T> {
-    pub fn new_from_db_wrapper(db_wrapper: DatabaseColumnScheduledDeleteWrapper) -> Self {
-        Self {
-            db_wrapper,
-            in_memory_index: InMemoryNumericIndex::default(),
-        }
-    }
-
-    pub fn into_in_memory_index(self) -> InMemoryNumericIndex<T> {
-        self.in_memory_index
-    }
-
-    pub fn db_wrapper(&self) -> &DatabaseColumnScheduledDeleteWrapper {
-        &self.db_wrapper
-    }
-
-    pub fn new(db: Arc<RwLock<DB>>, field: &str) -> Self {
+impl<T: Encodable + Numericable + Send + Sync + Default> MutableNumericIndex<T>
+where
+    Vec<T>: Blob,
+{
+    /// Open mutable numeric index from RocksDB storage
+    ///
+    /// Note: after opening, the data must be loaded into memory separately using [`load`].
+    pub fn open_rocksdb(db: Arc<RwLock<DB>>, field: &str) -> Self {
         let store_cf_name = numeric_index_storage_cf_name(field);
         let db_wrapper = DatabaseColumnScheduledDeleteWrapper::new(DatabaseColumnWrapper::new(
             db,
             &store_cf_name,
         ));
+        Self::open_rocksdb_db_wrapper(db_wrapper)
+    }
+
+    pub fn open_rocksdb_db_wrapper(db_wrapper: DatabaseColumnScheduledDeleteWrapper) -> Self {
         Self {
-            db_wrapper,
+            storage: Storage::RocksDb(db_wrapper),
             in_memory_index: InMemoryNumericIndex::default(),
         }
     }
 
-    pub fn load(&mut self) -> OperationResult<bool> {
-        if !self.db_wrapper.has_column_family()? {
+    /// Open mutable numeric index from Gridstore storage
+    ///
+    /// Note: after opening, the data must be loaded into memory separately using [`load`].
+    pub fn open_gridstore(path: PathBuf) -> OperationResult<Self> {
+        let options = default_gridstore_options::<T>();
+        let store = Gridstore::open_or_create(path, options).map_err(|err| {
+            OperationError::service_error(format!(
+                "failed to open mutable numeric index on gridstore: {err}"
+            ))
+        })?;
+        Ok(Self {
+            storage: Storage::Gridstore(Arc::new(RwLock::new(store))),
+            in_memory_index: InMemoryNumericIndex::default(),
+        })
+    }
+
+    /// Load storage
+    ///
+    /// Loads in-memory index from backing RocksDB or Gridstore storage.
+    pub(super) fn load(&mut self) -> OperationResult<bool> {
+        match self.storage {
+            Storage::RocksDb(_) => self.load_rocksdb(),
+            Storage::Gridstore(_) => Ok(self.load_gridstore()),
+        }
+    }
+
+    /// Load from RocksDB storage
+    ///
+    /// Loads in-memory index from RocksDB storage.
+    fn load_rocksdb(&mut self) -> OperationResult<bool> {
+        let Storage::RocksDb(db_wrapper) = &self.storage else {
```

Comment:
Why would we call `load_rocksdb` on a `Storage::Gridstore`?
This looks like it should be an error maybe.

Reply:
I agree, an error makes sense here. Thanks!

Reply:
I replaced all of them here: https://github.com/qdrant/qdrant/pull/6609/commits/b96b235c72b63fe111b38c840721bf0de75f111e